cmd/attest_test.go

```diff
@@ -0,0 +1,339 @@
+	defer client.CheckedClose(t, rwc)
+	ExternalTPM = rwc
+
+	inputFile := makeOutputFile(t, "attest")
+	outputFile := makeOutputFile(t, "attestout")
+
+	tests := []struct {
+		name           string
+		nonce          string
+		report         string
+		verifiedReport string
+		format         string
+	}{
+		{"Format:binary", "abcd", inputFile, outputFile, "binarypb"},
+		{"Format:textproto", "abcd", inputFile, outputFile, "textproto"},
+	}
+	for _, op := range tests {
+		t.Run(op.name, func(t *testing.T) {
+			attestArgs := []string{"attest", "--nonce", op.nonce, "--output", op.report, "--format", op.format}
+			RootCmd.SetArgs(attestArgs)
+			if err := RootCmd.Execute(); err != nil {
+				t.Error(err)
+			}
+
+			debugArgs := []string{"debugVerify", "--nonce", op.nonce, "--input", op.report, "--output", op.verifiedReport, "--format", op.format}
+			RootCmd.SetArgs(debugArgs)
+			if err := RootCmd.Execute(); err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestFormatFlagFail(t *testing.T) {
+	rwc := test.GetTPM(t)
+	defer client.CheckedClose(t, rwc)
+	ExternalTPM = rwc
+
+	inputFile := makeOutputFile(t, "attest")
+	outputFile := makeOutputFile(t, "attestout")
+
+	tests := []struct {
+		name           string
+		nonce          string
+		report         string
+		verifiedReport string
+		format_attest  string
+		format_debug   string
+	}{
+		{"Format:binary", "abcd", inputFile, outputFile, "binarypb", "textproto"},
+		{"Format:textproto", "abcd", inputFile, outputFile, "textproto", "binarypb"},
+		{"Format:textproto", "abcd", inputFile, outputFile, "textproto", "xyz"},
+	}
+	for _, op := range tests {
+		t.Run(op.name, func(t *testing.T) {
+			attestArgs := []string{"attest", "--nonce", op.nonce, "--output", op.report, "--format", op.format_attest}
+			RootCmd.SetArgs(attestArgs)
+			if err := RootCmd.Execute(); err != nil {
+				t.Error(err)
+			}
+
+			debugArgs := []string{"debugVerify", "--nonce", op.nonce, "--input", op.report, "--output", op.verifiedReport, "--format", op.format_debug}
+			RootCmd.SetArgs(debugArgs)
+			if err := RootCmd.Execute(); err == nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestAttestWithGCEAK(t *testing.T) {
+	rwc := test.GetTPM(t)
+	defer client.CheckedClose(t, rwc)
+	ExternalTPM = rwc
+
+	var template = map[string]tpm2.Public{
+		"RSA": GCEAKTemplateRSA(),
+		"ECC": GCEAKTemplateECC(),
+	}
+
+	tests := []struct {
+		name    string
+		nonce   []byte
+		keyAlgo string
+	}{
+		{"gceAK:RSA", []byte{1, 2, 3, 4}, "RSA"},
+		{"gceAK:ECC", []byte{1, 2, 3, 4}, "ECC"},
+	}
+	for _, op := range tests {
+		t.Run(op.name, func(t *testing.T) {
+			data, err := template[op.keyAlgo].Encode()
+			if err != nil {
+				t.Fatalf("failed to encode GCEAKTemplateRSA: %v", err)
+			}
+			k, err := gceAttestationKey(t, rwc, op.keyAlgo, data)
+			if err != nil {
+				t.Error(err)
+			}
+			defer k.Close()
+			attestOpts := client.AttestOpts{}
+			attestOpts.Nonce = op.nonce
+			attestation, err := k.Attest(attestOpts)
```

Comment:
We're testing `client` package functionality here. We should instead be doing calling the CLI:
```
if err := RootCmd.Execute(); err != nil {
```

This will also require setting up the fake MDS in this test, which the test should also verify.

Reply:
If we use CLI to test this ,then it will try to create a GCEAK from the client/attest.go and give the same error .
```
--- FAIL: TestAttestWithGCEAK/gceAK:RSA (0.08s)
        attest_test.go:284: failed to create attestation key: read error at index 29425667: decoding NV_ReadPublic response: handle 1, error code 0xb : the handle is not correct for the use
```

Reply:
I have changed the implementation of `TestAttestWithGCEAK()` and use attest CLI to test it.